subxt/src/backend/utils.rs

```diff
@@ -0,0 +1,161 @@
+    ///
+    /// The callback is invoked if a reconnection occurs.
+    pub fn with_resubscribe_callback(
+        stream: StreamOfResults<T>,
+        resubscribe: ResubscribeGetter<T>,
+    ) -> Self {
+        Self {
+            stream: Some(stream),
+            resubscribe: Some(resubscribe),
+            pending: None,
+        }
+    }
+
+    /// Returns the next item in the stream. This is just a wrapper around
+    /// [`StreamExt::next()`] so that you can avoid the extra import.
+    pub async fn next(&mut self) -> Option<Result<T, Error>> {
+        StreamExt::next(self).await
+    }
+}
+
+impl<T> std::marker::Unpin for RetrySubscription<T> {}
+
+impl<T> Stream for RetrySubscription<T> {
+    type Item = Result<T, Error>;
+
+    fn poll_next(
+        mut self: Pin<&mut Self>,
+        cx: &mut std::task::Context<'_>,
+    ) -> Poll<Option<Self::Item>> {
+        loop {
+            // Poll the stream.
+            let need_resubscribe = match self.stream.as_mut() {
+                Some(s) => match s.poll_next_unpin(cx) {
+                    Poll::Ready(Some(Err(e))) => {
+                        if e.is_disconnected_will_reconnect() {
+                            true
+                        } else {
+                            return Poll::Ready(Some(Err(e)));
+                        }
+                    }
+                    other => return other,
+                },
+                None => false,
+            };
+
+            if need_resubscribe {
+                self.stream = None;
+
+                let Some(f) = self.resubscribe.as_mut() else {
+                    tracing::error!("No callback configured for RetrySubscription that emitted Error::DisconnectedWillReconnect; This a bug please file an issue");
+                    return Poll::Ready(None);
+                };
+
+                self.pending = Some(f());
+            }
+
+            // Poll the resubscription.
+            let not_pending = if let Some(p) = self.pending.as_mut() {
+                if let Ok(stream) = ready!(p.poll_unpin(cx)) {
+                    self.stream = Some(stream);
+                    true
+                } else {
+                    false
+                }
+            } else {
+                false
+            };
+
+            if not_pending {
+                self.pending = None;
+            }
+        }
+    }
+}
+
+/// Retry a future with custom strategy.
+pub async fn retry_with_strategy<T, A>(
+    strategy: tokio_retry::strategy::ExponentialBackoff,
+    mut retry_future: A,
+) -> Result<T, Error>
+where
+    A: Action<Item = T, Error = Error>,
+{
+    RetryIf::spawn(
+        strategy,
+        || retry_future.run(),
+        |err: &Error| err.is_disconnected_will_reconnect(),
+    )
+    .await
+}
+
+/// Retry policy
+///
+/// Does nothing if a non-reconnecting rpc client hasn't been enabled.
+#[derive(Debug, Clone)]
+pub struct RetryPolicy(tokio_retry::strategy::ExponentialBackoff);
```

Comment:
Why do we actually need any sort of retry policy in Subxt? ie wouldn't the reconnecting RPC client handle this all? It would use a retry policy internally, and we would just get back Pending from it until it reconnects and is able to start processing whatever is waiting

Reply:
It's used for retrying individual futures in the backend impls

So, the reconnecting is manged by the reconnecting client but not individual method calls.
For the unstable backend it's not used for instance subscribing to blocks, runtime upgrades etc because it's managed by the driver itself.

For other methods and the legacy ones we want to retry calls until they are successful or returns an error which isn't Rpc::DisconnectWillReconnect

Reply:
> So, the reconnecting is manged by the reconnecting client but not individual method calls.

Would it be possible for the individual calls, upon seeing a "DisconnectedWillReconnect" error, to immediatelytry reconnecting by resubscribing or whatever? And then the RPC client will return `Poll::Pending` until it's ready (ie it actually reconnects). And it could also be up to the RPC client how many times to try reconnecting etc; if it gives up then it emits some `Error::ConnecitonClosed` or whatever it is and then the backend gives up and errors back to the user :)

In that way, the backend/subxt doesn't need to think at all about retry policies etc and it's entirely up to the RPC client itself to manage.

Reply:
I'm a little confused by this.

So, the reconnecting client supporst to resubscribe/retry method calls and reconnecting but only reconnecting is enabled now in subxt because that we don't want to resubscribe to stuff like `authorSubmitAndWatch` because it has side-effects.

Instead it's up to backend impl to decide whether retry or not but the reconnecting rpc client reconnects "automatically".

Indeed, we could return Poll::Pending in the reconnecting client instead of having this code: https://github.com/paritytech/subxt/pull/1505/files#diff-fc19ff73c9c406c6d7e3da3c3dda26c8cc6f2727ae358037d1be63b6cfdfdfb2R96-R110. Yepp, that would be neat

Reply:
I had a look at this and it's a quite hard to implement such a thing but I realized we already have this functionality in the reconnecting-rpc-client.

So I enabled "retry" for rpc calls/subscriptions from the reconnecting rpc client but turned off "resubscribing" which achieves the same thing.

The downside is that it's quite hard to implement custom retry behavior rather than the global policy but I fine with that for now.

We could enable that with a HashMap and just turn it off for certain methods....

Reply:
I'm closing this discussion my last comment isn't accurate anymore.

TLDR; subxt is handling retry logic for subscriptions, method calls and the reconnecting client just reconnects.